### PR TITLE
Add quality requirements for AppCat

### PR DESCRIPTION
**Pull request by Kidswiss**
Adds various quality requirements to the AppCat Framework.

Format is inspired by https://github.com/appuio/appuio-io-docs/

### CODE REVIEW

File: docs/modules/ROOT/pages/reference/quality-usability.adoc [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+Webbrowser
+
+Response::
+All logs of the service are available
+
+Response measure::
+The service's logs are available up to the normal logging retention.
+
+Rationale::
+To give the users more flexibility with debugging issues and using the services, we will provide access to the service's logs.
```
Comment (ccremer) [inline docs/modules/ROOT/pages/reference/quality-usability.adoc:93]: I doubt we can make this part of the first iteration.
I would exclude it for the time being, unless it's already defined and I missed it?

**Review by ccremer — COMMENTED** [review]
Overall a good starting point :+1: 
This PR has gotten very large and I didn't fully review it yet.
Might consider splitting it up in some way.

Reading through it could roughly group the requirements into 2 categories:
* The ones targeting the AppCat as a Framework itself to establish a base standard. (e.g. some in usability or regarding monitoring)
* The ones that are try to dictate how services themselves should behave or be engineered. (e.g. security, performance)

I think it's better to only list the ones directly applicable for the framework. And then each service should be treated independently including things like security or performance aspects.

File: docs/modules/ROOT/pages/reference/quality-functional.adoc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+AppCatalog enabled Kubernetes cluster
+
+Artifact::
+AppCatalog Service
+
+Response::
+Service created
+
+Response measure::
+Within a few minutes, see service for specific definition.
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:21]: We can do that, but 30 minutes feel like a bit long. Maybe we also have to differentiate between single-instances and cluster? Single-instances will probably be within 5 minutes, and clusters could take up to 15 minutes.

File: docs/modules/ROOT/pages/reference/quality-functional.adoc [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+Creating a service requires service specific actions, depending on the service they time they take to provision may vary.
+
+[NOTE]
+====
+The provided APIs can vary according to the environment.
+====
+
+== Configuration of Services Is a Declarative Approach
+Source::
+User
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:34]: Yes, makes sense to put it there.

File: docs/modules/ROOT/pages/reference/quality-functional.adoc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+Kubernetes and https://kb.vsn.ch[VSHN Documentation]
+
+Response::
+The service attempts self-healing
+
+Response measure::
+If the self-healing doesn't work, or not automation is provided for the given issue, then runbooks should cover these cases.
+
+Rationale::
+Each engineer should be able to debug an AppCatalog service during operations or OnCall.
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:75]: The remark about self-healing was about k8s self-healing. But yes my ultimate goal was "no alert without runbook".

File: docs/modules/ROOT/pages/reference/quality-functional.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (ccremer) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:21]: Thinking about it again, I'm not sure it even makes sense to put a deadline on provisioning. Considering we now only go for "best-effort" and SLA-backed support options, there's no reason that the service has to be running after x minutes for the best-effort case. And in case of guaranteed availability, I guess this should somehow be alert-based.

[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:21]: I'm not sure if an alert would make sense here. It could lead to some false positives, probably if a customer orders something while cluster maintenance is ongoing.

Also, how would that be actionable? If the provisioning takes longer due to infrastructure or longer image pull times, then this is rather a symptom  of another issue and not an issue itself.

File: docs/modules/ROOT/pages/reference/quality-portability.adoc [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+Kubectl or Service Broker API
+
+Response::
+The services are the same for all any AppCatalog enabled Kubernetes cluster zones.
+
+Response measure::
+The exception are resold Cloud services like RDS or CloudSQL.
+
+Rationale::
+The AppCatalog service does not care on what k8s they are running.
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-portability.adoc:23]: Maybe I should rephrase it a bit, instead of Kubernetes we could say cloud. 

> Postgresql on Cloudscale is entirely different than on AWS.

Do you refer to the difference between Postgres via helm on k8s and RDS on AWS?

File: docs/modules/ROOT/pages/reference/quality-portability.adoc [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+K8up and restic
+
+Response::
+The backup is restored
+
+Response measure::
+We don't want to vendor lock the users into our services. The customer should be able to take their data away by themselves.
+
+Rationale::
+The AppCatalog service does not care on what k8s they are running.
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-portability.adoc:45]: Even if it's a black box, the customer can still request a restore of the data. This also implies that each service has a backup that can be restored.

File: docs/modules/ROOT/pages/reference/quality-security.adoc [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+Kubectl or Service Broker API
+
+Response::
+The service only accepts TLS encrypted connections
+
+Response measure::
+The connection details contain all necessary parameters to connect to the service securely.
+
+Rationale::
+On a shared environment, the encryption of in-transit data is an important security aspect.
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-security.adoc:23]: I can rewrite it a bit so that it's not a hard requirement to support TLS, but available for feasible use-cases.

File: docs/modules/ROOT/pages/reference/quality-security.adoc [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+Crossplane and Helm
+
+Response::
+The default service configuration is hardened
+
+Response measure::
+Security best practices are applied to each AppCatalog Service. Authentication is enabled by default (if available).
+
+Rationale::
+In a shared cloud environment, it is crucial to secure the services, so no other people have access to it.
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-security.adoc:46]: Not necessarily.

For example, Redis doesn't have authentication by default. So our hardening it to enable it. This isn't necessarily coupled to TLS or not.

File: docs/modules/ROOT/pages/reference/quality-functional.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (tobru) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:21]: Service provisioning time isn't part of the SLA, so there is nothing we have to guarantee. But IMHO tracking the time how long it takes to provision a service should be important for us. I'd say let's go for 30 minutes and learn how all the services behave when we have them. We can then revisit this number.

File: docs/modules/ROOT/pages/reference/quality-performance.adoc [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+https://kb.vshn.ch[VSHN Documentation]
+
+Response::
+Performance is within the defined requirements.
+
+Response measure::
+The service documentation describe what performance can be expected of it.
+
+Rationale::
+The service performance is specific to each service and has to be described separately.
```
[the remark this answers is not in the document]
Comment (tobru) [reply docs/modules/ROOT/pages/reference/quality-performance.adoc:24]: Thinking about that one again, I'd vote to remove it for the time being. Performance isn't something we can currently guarantee and describe properly.

File: docs/modules/ROOT/pages/reference/quality-portability.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (tobru) [reply docs/modules/ROOT/pages/reference/quality-portability.adoc:23]: IMHO this requirement targets how the service is provisioned. Ordering a Postgres instance on AWS should be the same as ordering it on cloudscale.ch.

File: docs/modules/ROOT/pages/reference/quality-functional.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (ccremer) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:21]: From a Software architecture textbook PoV, it's the quality requirements that make up the software architecture. In this case however, I fail to see the impact this time deadline has on the overall architecture. What happens if the provisioning fails in x minutes and how do we ensure they stay under the limit? For me service provisioning is something a là "it takes as long as it needs". In case of abnormal conditions this requirement hardly covers such cases anyway.
If we just track the time for dashboard purposes, I don't think we need to formulate a quality requirement here at all since for me, these are very strong and formal. If we just go for the dashboard purpose, a simple functional requirement (feature) like "internal dashboard for service provisioning time" somewhere else would make more sense.

File: docs/modules/ROOT/pages/reference/quality-functional.adoc [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+Rationale::
+Creating a service requires service specific actions, depending on the service the time it takes to provision may vary.
+
+[NOTE]
+====
+- The provided APIs can vary according to the environment.
+- The service is provisioned asynchronously and its readiness status has to be polled regularly.
+====
+
+== Configuration of Services Is a Declarative Approach
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:32]: Should be capitalized according to: https://capitalizemytitle.com/#capitalizationrules

File: docs/modules/ROOT/pages/reference/quality-portability.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (ccremer) [reply docs/modules/ROOT/pages/reference/quality-portability.adoc:45]: Then I would rephrase the requirement for the time being
```suggestion
== Backups Can Be Exported
Source::
Service engineer

Stimulus::
Export an existing backup

Environment::
AppCatalog enabled Kubernetes cluster

Artifact::
K8up and restic

Response::
The backup is in a format that can be exported.

Response measure::
The backup data can be inspected and used to fully restore a service instance outside of AppCat with conventional tools.

Rationale::
Vendor lock-in users into the service instances or AppCat framework must be avoided.
The users should be able to take their data away before canceling their service instances.
```

File: docs/modules/ROOT/pages/reference/quality-usability.adoc (hunk 1 above)
Comment (ccremer) [reply docs/modules/ROOT/pages/reference/quality-usability.adoc:93]: Yes, but there's a lot of open questions how we can grant access to the namespaces and their logs without giving potentially destructive access to service namespaces that are more or less treated as blackboxes for users.

Besides, this could be fine with APPUiO Cloud, but having this requirement here basically means that it applies also to each existing and future private managed cluster, where definitely not all clusters have even central logging enabled. So I would remove it for the time being as a general requirement, and list this feature as exclusive for APPUiO Cloud for now. At least until we know if and how we aim to achieve the same for private clusters. 

The difference is that with a clear dedicated requirement like this we commit ourselves that each AppCat environment fulfills it, even in cases where OpenShift logging isn't available and we would have to engineer something ourselves. However, if the central logging feature is outsourced to the platform (OpenShift), then we can argue that this feature availability depends on the platform.

File: docs/modules/ROOT/pages/reference/quality-functional.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (tobru) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:21]: For me the time it takes until a service is provisioned is a quality requirement. If it takes ages, I might choose a different service provider. We should take appropriate measures to have a service provisioned in due time, and we should make sure we meet this time constraint, otherwise the quality requirement is violated, and the service needs to be investigated.

File: docs/modules/ROOT/pages/reference/quality-functional.adoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (tobru) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:32]: Well...

File: docs/modules/ROOT/pages/reference/quality-portability.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (tobru) [reply docs/modules/ROOT/pages/reference/quality-portability.adoc:23]: I could very well imagine such a case, yes.

File: docs/modules/ROOT/pages/reference/quality-usability.adoc (hunk 1 above)
Comment (tobru) [reply docs/modules/ROOT/pages/reference/quality-usability.adoc:93]: We should rephrase the requirement to something like "User Has Access To Logs of Service", which doesn't mention how and if these logs are stored for a long time. Technically speaking an "oc -n my-black-box-service logs mydb" should be possible.

From a users' perspective having access to logs is a USP, therefore we should keep this requirement. If we see over time that this isn't feasible, we can always revisit this requirement and adapt.

File: docs/modules/ROOT/pages/reference/quality-functional.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (ccremer) [reply docs/modules/ROOT/pages/reference/quality-functional.adoc:21]: I see. Then I would suggest something like this:
```suggestion
Response::
Service ready

Response measure::
Service ready depending on service architecture in normal conditions:
- Standalone: within 10 minutes
- Replicated: within 20 minutes
- Clustered: within 30 minutes
- Cloud instance: within 20 minutes
```

File: docs/modules/ROOT/pages/reference/quality-portability.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-portability.adoc:45]: Thx for the suggestion.

File: docs/modules/ROOT/pages/reference/quality-usability.adoc (hunk 1 above)
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-usability.adoc:93]: I see @ccremer's point.

My suggestion, I rephrase it for the time being, that this only applies to the environment "APPUiO Cloud". We can re-visit it in a later stage ans extend as needed.

**Comment by ccremer** [comment]
I feel these discussions are somewhat inefficient but very important. Let's make a call to discuss the details and then do some pair programming to get this done.

File: docs/modules/ROOT/pages/reference/quality-reliability.adoc [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+Crossplane and Helm
+
+Response::
+New versions of services are made available
+
+Response measure::
+If a new version for a service is released, this version will be provided in a timely manner.
+
+Rationale::
+Up-to-date services contain bugfixes and security patches that improve the service's reliability.
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply docs/modules/ROOT/pages/reference/quality-reliability.adoc:70]: I'll rephrase it, that the customer can expect their services to be updated. We'll add renovate and further automation later.

**Comment by Kidswiss** [comment]
I've rephrased the points that were still open. We can take that as a starting point for a call.

**Comment by ccremer** [comment]
@Kidswiss and I had a call to flesh out duplicates and things that are better left documented in the future service development guide. This PR got smaller and more structured.

**Review by tobru — APPROVED** [review]

**Review by ccremer — APPROVED** [review]
